### PR TITLE
Add gradient norm logger with PyTorch Lightning integration, tests, and hook profiler

### PR DESCRIPTION
**Pull request by Jayantparashar10**
## Description

This PR adds gradient norm logging and PyTorch Lightning integration to Visdom.

Right now there's no way to monitor gradient norms without writing the math yourself and calling `viz.line()` manually. If you're using Lightning, there's no logger at all, so you end up duplicating the same boilerplate in every project. This adds proper support for both.

New files:

- `py/visdom/grad_norm.py` contains two utility functions: `compute_grad_norm()` for the global Lp norm (same formula as `torch.nn.utils.clip_grad_norm_`) and `compute_layer_grad_norms()` for per-parameter norms with named keys.

- `py/visdom/lightning_logger.py` has the Lightning integration. `VisdomLogger` is a full `Logger` subclass so you just pass it to `Trainer` and every `self.log()` call in your `LightningModule` automatically gets plotted in Visdom. Each run gets its own versioned environment so nothing gets overwritten. `GradientNormCallback` registers autograd tensor hooks during training so gradient norms are captured and logged without touching your model code at all.

- `example/pytorch_mnist_demo.py` is a manual MNIST training loop using `viz.line()` directly, as a baseline.

- `example/demo_lightning_mnist_grad.py` is the same training setup rewritten with Lightning, showing what the workflow looks like when you use `VisdomLogger` and `GradientNormCallback` together.

- `example/hook_overhead_profile.py` runs a quick benchmark across five hook scenarios (no hook, forward_pre, forward, backward, grad norm) to show how much overhead each adds.

- `tests/test_loggers.py` has 53 unit tests covering the norm math, the `log_gradient_norm` method, logger behavior, and the callback lifecycle. Everything is mocked so no live server is needed to run them.

Modified:

- `py/visdom/__init__.py` gets a new `log_gradient_norm()` method on the `Visdom` client and lazy imports so `from visdom import VisdomLogger, GradientNormCallback` works out of the box without breaking anything existing.

## Motivation and Context

Visdom has no built-in gradient norm logging and no Lightning integration, so users end up writing `viz.line()` calls manually for every metric. `VisdomLogger` works the same way TensorBoard or W&B loggers do, versioned environments per run so previous runs are not overwritten. `GradientNormCallback` hooks into autograd so grad norms are captured during the backward pass with no changes needed in the model. For users not using Lightning, `log_gradient_norm()` is a simple one-liner on the existing `Visdom` client.

## How Has This Been Tested?

Ran the full test suite:
```
python -m pytest tests/test_loggers.py -q --rootdir=. --import-mode=importlib
```
Tests cover the grad norm math, the `log_gradient_norm` method, logger metric and hyperparam logging, and the callback's hook lifecycle, `log_every` throttle, and `per_layer` mode.

Also ran the end-to-end demo:
```
python -m visdom.server
python example/demo_lightning_mnist_grad.py
```
Opened `http://localhost:8097`, switched to the `mnist_lightning-v*` environment, and confirmed all windows showed up: `train_loss`, `val_loss`, `val_acc`, `grad_norm`, per-layer `grad_norm/model/*`, `grad_hook_ms`, and `hparams`. Training ran for 5 epochs and finished at `val_acc` 0.992.

Hook benchmark:
```
python example/hook_overhead_profile.py
```
The gradient norm hook adds less than 1% overhead on average.

Installed from source (`pip install -e .`) and verified no regressions in basic manual logging.

Tested on Python 3.12, PyTorch 2.10.0, lightning 2.x.

## Screenshots (if appropriate):

Visdom dashboard after running `demo_lightning_mnist_grad.py` (env `mnist_lightning-v20260303_012236`):

image

image

- Per-layer grad norm charts for `conv1/weight`, `conv2/weight`, `fc1/weight`, etc., all decaying over roughly 4700 steps
- Global `grad_norm` chart following the same trend
- `val_loss` dropping from around 0.05 to 0.025 over 5 epochs, `val_acc` reaching 0.992
- `train_loss` dropping from around 0.15 to 0.03
- `grad_hook_ms` staying near zero, confirming negligible hook overhead
- `hparams` table with `lr=0.001`

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactor or cleanup (changes to existing code for improved readability or performance)

## Checklist:
- [ ] I adapted the version number under `py/visdom/VERSION` according to [Semantic Versioning](https://semver.org/)
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: py/visdom/__init__.py [hunk 1]
```diff
@@ -2592,3 +2596,137 @@ def graph(
...
+                    X=np.array(list(layer_norms.values())),
+                    win=win + "_layers",
+                    env=resolved_env,
+                    opts=dict(
+                        title=win + " (per layer)",
+                        rownames=list(layer_norms.keys()),
+                    ),
+                )
+
+        return total_norm
```
[the remark this answers is not in the document]
Comment (Jayantparashar10) [reply py/visdom/__init__.py:2710]: Fixed - `log_gradient_norm` has been added to ```__init__.pyi``` with the correct signature.

File: py/visdom/__init__.py [hunk 2]
```diff
@@ -2592,3 +2596,137 @@ def graph(
         return self._send(
             {"data": data, "win": win, "eid": env, "opts": opts}, endpoint="events"
         )
+
+    def log_gradient_norm(
```
[the remark this answers is not in the document]
Comment (Jayantparashar10) [reply py/visdom/__init__.py:2600]: The latest version of PyTorch is 2.10.0

File: py/visdom/__init__.pyi [hunk 3]
```diff
@@ -210,3 +210,13 @@ class Visdom:
...
+    def log_gradient_norm(
+        self,
+        model: Any,
+        step: int,
+        win: str = ...,
+        env: _OptStr = ...,
+        opts: _OptOps = ...,
+        norm_type: float = ...,
+        per_layer: bool = ...,
+    ) -> float: ...
```
[the remark this answers is not in the document]
Comment (Jayantparashar10) [reply py/visdom/__init__.pyi:222]: Added in the `py/visdom/__init__.pyi`

File: py/visdom/lightning_logger.py [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+                    Y=np.array([scalar]),
+                    X=np.array([x_val]),
+                    win=win,
+                    env=self._env,
+                    update="append" if already else None,
+                    opts=opts,
+                )
+                self._created_windows.add(win)
+            except Exception as exc:  # pragma: no cover
+                warnings.warn(f"[VisdomLogger] Failed to log '{key}': {exc}")
```
[the remark this answers is not in the document]
Comment (Jayantparashar10) [reply py/visdom/lightning_logger.py:266]: Added in the code 
```python 
try:
       self.experiment.line(
        Y=np.array([scalar]),
        X=np.array([x_val]),
        win=win,
        env=self._env,
        update="append" if already else None,
        opts=opts,
        )
        self._created_windows.add(win)
except Exception as exc:  # pragma: no cover
             if self._fail_on_no_connection:
                    raise
            warnings.warn(f"[VisdomLogger] Failed to log '{key}': {exc}")
 ```
